components/core/src/clp_s/JsonParser.cpp

```diff
@@ -520,6 +521,317 @@ bool JsonParser::parse() {
     return true;
 }
 
+NodeType get_archive_node_type(
+        clp::ffi::SchemaTreeNode const& node,
+        std::pair<clp::ffi::SchemaTreeNode::id_t, std::optional<clp::ffi::Value>> p
+) {
+    auto const node_type = node.get_type();
+    // figure out what type the node is in archive node type
+    NodeType archiveNodeType;
+    switch (node_type) {
+        case clp::ffi::SchemaTreeNode::Type::Int:
+            archiveNodeType = NodeType::Integer;
+            break;
+        case clp::ffi::SchemaTreeNode::Type::Float:
+            archiveNodeType = NodeType::Float;
+            break;
+        case clp::ffi::SchemaTreeNode::Type::Bool:
+            archiveNodeType = NodeType::Boolean;
+            break;
+        case clp::ffi::SchemaTreeNode::Type::UnstructuredArray:
+            archiveNodeType = NodeType::UnstructuredArray;
+            break;
+        case clp::ffi::SchemaTreeNode::Type::Str:
+            // std::cerr << "In str\n";
+            if (p.second.value().is<std::string>()) {
+                // maybe special case for date string
+                archiveNodeType = NodeType::VarString;
+            } else {
+                archiveNodeType = NodeType::ClpString;
+            }
+            break;
+        case clp::ffi::SchemaTreeNode::Type::Obj:
+            // std::cerr << "In obj\n";
+            if (p.second.has_value()) {
+                if (p.second.value().is_null()) {
+                    // std::cout << "Found Null\n";
+                    archiveNodeType = NodeType::NullValue;
+                } else {
+                    archiveNodeType = NodeType::Object;
+                }
+            } else {
+                archiveNodeType = NodeType::Object;
+            }
+            break;
+        default:
+            archiveNodeType = NodeType::Unknown;
+            break;
+    }
+    return archiveNodeType;
+}
+
+//
+int JsonParser::get_archive_node_id(
+        std::map<std::tuple<int, NodeType>, int>& cache,
+        int irNodeID,
+        NodeType archiveNodeType,
+        clp::ffi::SchemaTree const& irTree
+) {
+    std::tuple<int, NodeType> key(irNodeID, archiveNodeType);
+    if (cache.find(key) != cache.end()) {
+        return cache[key];
+    }
+    auto& currNode = irTree.get_node(irNodeID);
+    int parent_node_id;
+    // Found the root
+    if (currNode.get_parent_id() == 0) {
+        parent_node_id = 0;
+    } else {
+        parent_node_id
+                = get_archive_node_id(cache, currNode.get_parent_id(), NodeType::Object, irTree);
```

Comment:
```suggestion
    // Note: assuming zero here will be unsafe once kv pair ir supports internal metadata fields
    int32_t parent_node_id{0};
    if (0 != curr_node.get_parent_id()) {
        parent_node_id
                = get_archive_node_id(cache, curr_node.get_parent_id(), NodeType::Object, ir_tree);
```
More concise. Also literals always go on the left for comparisons in an if statement -- it helps avoid accidental assignments like `if (a = 0) {}`.

One thing I'm worried about is assuming that the root of the object subtree has ID 0 in the archive. This is true for now, but won't be true in general (e.g. extra internal fields can throw off these IDs, or ir being ingested alongside other logs in the same archive could lead to a different type of structure claiming ID 0 as its root).

Ideally we would explicitly track the root of the object subtree when we first add it to the archive while ingesting the IR, then use that as parent_node_id's default value here.

Reply:
I modified this to check that current node id and the parent node _id aren't the same. If they are the same then we don't recurse.